src/main/java/frc/robot/subsystems/conveyor/Conveyor.java

```diff
@@ -0,0 +1,198 @@
+        if (INSTANCE == null) {
+            INSTANCE = new Conveyor();
+        }
+        return INSTANCE;
+    }
+
+    /**
+     * @return the color sensor value as a DriverStation.Alliance enum.
+     */
+    public DriverStation.Alliance getColor() {
+        Color color = colorSensorIntake.getColor();
+        ColorMatchResult result = match.matchClosestColor(color);
+        Color resultColor = result.color;
+        SmartDashboard.putBoolean("match", resultColor == Color.kRed);
+
+        if (resultColor == Constants.Conveyor.RED) {
+            return DriverStation.Alliance.Red;
+        } else if (resultColor == Constants.Conveyor.BLUE) {
+            return DriverStation.Alliance.Blue;
+        } else {
+            return DriverStation.Alliance.Invalid;
+        }
+    }
+
+    /**
+     * @return the amount of balls inside the cargo.
+     */
+    public int getCargoCount() {
+        return position.size();
+    }
+
+    /**
+     * Set motor power.
+     *
+     * @param power the power to the motor [-1, 1].
+     */
+    public void setPower(double power) {
+        motor.set(power);
+    }
+
+    /**
+     * open the "flap" - solenoid
+     */
+    public void openFlap() {
+        flap.set(false);
+    }
+
+    /**
+     * closes the "flap" - solenoid
+     */
+    public void closeFlap() {
+        flap.set(true);
+    }
+
+    /**
+     * toggles the "flap" - solenoid
+     */
+    public void toggle() {
+        flap.toggle();
+    }
+
+    private void updateActualBallPositions() {
+        var colorIntake = getColor();
+        boolean currentPostFlapBeamInput = postFlapBeam.get();
+        SmartDashboard.putString("alliance", colorIntake.name());
+        if (currentPostFlapBeamInput && !lastPostFlapBeamInput && motor.getMotorOutputPercent() > 0) {
+            position.removeFirst();
+        }
+        if (colorIntake != lastSeenColor) {
+            if (motor.getMotorOutputPercent() < 0) {
+                position.removeLast();
+            } else if (motor.getMotorOutputPercent() > 0) {
+                position.add(colorIntake.name());
+            }
+        }
+        lastPostFlapBeamInput = currentPostFlapBeamInput;
+        lastSeenColor = colorIntake;
+        SmartDashboard.putString("lastSeen", lastSeenColor.name());
+    }
+
+    private Queue getQueue() {
+        /*
+        Case #1  queue = [None, None]
+        Case #2  queue = [None, Alliance]
+        Case #3  queue = [Alliance, None]
+        Case #4  queue = [Alliance, Alliance]
+        Case #5  queue = [Alliance, Opponent]
+        Case #6  queue = [None, Opponent]
+        Case #7  queue = [Opponent, None]
+        Case #8  queue = [Opponent, Opponent]
+        Case #9  queue = [Opponent, Alliance]
+         */
+        var last = position.getLast();
+        var first = position.getFirst();
+        var invalid = DriverStation.Alliance.Invalid.name();
+        var alliance = DriverStation.getAlliance().name();

```

Comment:
1. invalid what? alliance what?
2. `invalid` and `alliance` can be declared at the subsystem's constructor, no need to create them repeatedly.
3. Can't `invalid` be replaced with `Constants.Conveyor.NONE`?